File: sdks/ruby/lib/blok/config/server_config.rb
# frozen_string_literal: true

module Blok
  module Config
    # ServerConfig holds runtime server configuration, loadable from environment variables.
    class ServerConfig
      # Selects which transport(s) to start.
      module Transport
        HTTP = :http
        GRPC = :grpc
        BOTH = :both
      end

      attr_accessor :port, :host, :version, :grpc_port, :transport, :log_level, :enable_cors

      # @param port [Integer] HTTP port (default: 9006 — matches DEFAULT_PORTS.ruby)
      # @param host [String] Bind host (default: "0.0.0.0")
      # @param version [String] Runtime version (default: Blok::VERSION)
      # @param grpc_port [Integer] gRPC port (default: 10006 — matches DEFAULT_GRPC_PORTS.ruby = HTTP+1000)
      # @param transport [Symbol] One of :http | :grpc | :both (default: :http)
      # @param log_level [Symbol] Minimum log level (default: :info)
      # @param enable_cors [Boolean] Enable CORS headers (default: false)
      def initialize(port: 9006, host: "0.0.0.0", version: nil,
                     grpc_port: 10006, transport: Transport::HTTP,
                     log_level: Logging::LogLevel::INFO, enable_cors: false)
        @port        = port
        @host        = host
        @version     = version || Blok::VERSION
        @grpc_port   = grpc_port
        @transport   = transport
        @log_level   = log_level
        @enable_cors = enable_cors
      end

      # Load configuration from environment variables with sensible defaults.
      #
      # Environment variables:
      #   PORT           - HTTP port (default: 9006)
      #   HOST           - Bind host (default: "0.0.0.0")
      #   VERSION        - Runtime version (default: Blok::VERSION)
      #   GRPC_PORT      - gRPC port (default: 10006)
      #   BLOK_TRANSPORT - "http" | "grpc" | "both" (default: "http"; invalid -> "http")
      #   LOG_LEVEL      - Minimum log level: DEBUG, INFO, WARN, ERROR (default: INFO)
      #   ENABLE_CORS    - Enable CORS: "true" or "false" (default: false)
      #
      # @return [ServerConfig]
      def self.from_env
        new(
          port:        Integer(ENV.fetch("PORT", "9006")),
          host:        ENV.fetch("HOST", "0.0.0.0"),
          version:     ENV.fetch("VERSION", Blok::VERSION),
          grpc_port:   Integer(ENV.fetch("GRPC_PORT", "10006")),
          transport:   parse_transport(ENV.fetch("BLOK_TRANSPORT", "http")),
          log_level:   Logging::LogLevel.parse(ENV.fetch("LOG_LEVEL", "INFO")),
          enable_cors: ENV.fetch("ENABLE_CORS", "false") == "true"
        )
      end

      # Parse a transport string into a Transport symbol. Invalid values fall
      # back to {Transport::HTTP} so a typo never crashes the boot.
      def self.parse_transport(value)
        case value.to_s.downcase
        when "grpc" then Transport::GRPC
        when "both" then Transport::BOTH
        else Transport::HTTP
        end
      end

      # Return the bind address as "host:port".
      # @return [String]
      def address
        "#{@host}:#{@port}"
      end
    end
  end
end

File: sdks/python3/blok/config/server_config.py
from __future__ import annotations
import os


class ServerConfig:
    """Configuration for the blok HTTP server.

    All values can be overridden via environment variables.

    Environment variables:
        PORT: HTTP port (default: 9007)
        HOST: Bind address (default: 0.0.0.0)
        VERSION: Runtime version (default: 1.0.0)
        GRPC_PORT: gRPC port (default: 10007 — matches runner's
                   DEFAULT_GRPC_PORTS.python3 = HTTP+1000)
        BLOK_TRANSPORT: Transport selector. One of "http", "grpc", "both"
                        (default: "http"). When "grpc" or "both" the SDK
                        starts a gRPC server on GRPC_PORT.
        READ_TIMEOUT: Read timeout in seconds (default: 30)
        WRITE_TIMEOUT: Write timeout in seconds (default: 30)
        SHUTDOWN_TIMEOUT: Shutdown timeout in seconds (default: 10)
        LOG_LEVEL: Minimum log level: DEBUG, INFO, WARN, ERROR (default: INFO)
        ENABLE_CORS: Enable CORS: true/false (default: false)
    """

    def __init__(
        self,
        port: int = 9007,
        host: str = "0.0.0.0",
        version: str = "1.0.0",
        grpc_port: int = 10007,
        transport: str = "http",
        read_timeout_sec: int = 30,
        write_timeout_sec: int = 30,
        shutdown_timeout_sec: int = 10,
        log_level: str = "INFO",
        enable_cors: bool = False,
    ):
        self.port = port
        self.host = host
        self.version = version
        self.grpc_port = grpc_port
        self.transport = transport
        self.read_timeout_sec = read_timeout_sec
        self.write_timeout_sec = write_timeout_sec
        self.shutdown_timeout_sec = shutdown_timeout_sec
        self.log_level = log_level
        self.enable_cors = enable_cors

    @property
    def address(self) -> str:
        return f"{self.host}:{self.port}"

    @classmethod
    def from_env(cls) -> ServerConfig:
        """Load configuration from environment variables."""
        cfg = cls()

        if v := os.environ.get("PORT"):
            try:
                port = int(v)
                if port > 0:
                    cfg.port = port
            except ValueError:
                pass

        if v := os.environ.get("HOST"):
            cfg.host = v

        if v := os.environ.get("VERSION"):
            cfg.version = v

        if v := os.environ.get("GRPC_PORT"):
            try:
                gport = int(v)
                if gport > 0:
                    cfg.grpc_port = gport
            except ValueError:
                pass

        if v := os.environ.get("BLOK_TRANSPORT"):
            if v in ("http", "grpc", "both"):
                cfg.transport = v

        if v := os.environ.get("READ_TIMEOUT"):
            try:
                t = int(v)
                if t > 0:
                    cfg.read_timeout_sec = t
            except ValueError:
                pass

        if v := os.environ.get("WRITE_TIMEOUT"):
            try:
                t = int(v)
                if t > 0:
                    cfg.write_timeout_sec = t
            except ValueError:
                pass

        if v := os.environ.get("SHUTDOWN_TIMEOUT"):
            try:
                t = int(v)
                if t > 0:
                    cfg.shutdown_timeout_sec = t
            except ValueError:
                pass

        if v := os.environ.get("LOG_LEVEL"):
            if v in ("DEBUG", "INFO", "WARN", "ERROR"):
                cfg.log_level = v

        if os.environ.get("ENABLE_CORS") in ("true", "1"):
            cfg.enable_cors = True

        return cfg
